# Path of Exile game stats data eda

dates: 2022.05.17 ~ 

This file contains an analysis processing of the game 'Path of Exile' data.

###### Origins of datas from here: 
poe_stats : https://www.kaggle.com/datasets/gagazet/path-of-exile-league-statistic

Data contains stats of 59000 players, from 4th August of 2017 and before now.

---

특정 유저들을 골라내는 위주로 분석을 진행할 노트북입니다. 이름을 이용하거나, 트위치 아이디를 이용하는 방식입니다.



# Dataset and libraries 

## data and libraries import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
# import wquantiles

# from sklearn.utils import resample
# from scipy import stats
# from scipy.stats import trim_mean
# from statsmodels import robust
# import statsmodels.api as sm
# import statsmodels.formula.api as smf
# from statsmodels.stats import power

df = pd.read_csv('poe_stats_2revisedrank.csv')

In [2]:
df.head(10)

rank   dead  online                    name  level        class  \
0     1  False   False      Tzn_NecroIsFineNow    100  Necromancer   
1     1  False   False        RaizNeverFirstQT    100  Necromancer   
2     1  False   False  GucciStreamerAdvantage    100  Necromancer   
3     1  False   False             ChiroxPrime    100       Slayer   
4     2  False   False     Cool_NecroIsFineNow    100      Deadeye   
5     2  False   False  SteelStreamerAdvantage    100      Deadeye   
6     2  False   False   Grom_WhyGodThreeCmWhy    100    Gladiator   
7     2  False   False          beobio_capitan    100   Inquisitor   
8     3  False   False    Rami_The_Negromaniac    100  Necromancer   
9     3  False   False           FutonBlewAway    100       Raider   

                                                  id  experience  \
0  3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444   
1  8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444   
2  c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444   
3  c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444   
4  24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444   
5  51acb8e6fac3cfaa7ff848151dad01b0a6f350e7525543...  4250334444   
6  5d8b80b2bdd30f3dba1036a6791187562758935e8d4723...  4250334444   
7  65c58f9b186e1fbe9f927484f3f178cdb9ee0dc2427289...  4250334444   
8  00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...  4250334444   
9  46e56572ff6e403f45fac597240e13e9af225f06ca4e52...  4250334444   

         account  challenges        twitch              ladder  special_class  \
0         TheTzn          40        thetzn           Harbinger              1   
1         RaizQT          22        raizqt    SSF Harbinger HC              1   
2    GucciPradas          35   guccipradas  Hardcore Harbinger              1   
3       Chiroxun          26           NaN       SSF Harbinger              1   
4       cooltail          22           NaN           Harbinger              1   
5      Steelmage          37  steelmage202  Hardcore Harbinger              1   
6     rocklegend          22       gromick    SSF Harbinger HC              1   
7  TishaVosemsot          40        beobio       SSF Harbinger              1   
8       rami1337          40      rami1337           Harbinger              1   
9        pleaGRC          40           NaN       SSF Harbinger              1   

   stat_Str  stat_Int  stat_Dex  stats_num  
0         0         1         0          1  
1         0         1         0          1  
2         0         1         0          1  
3         1         0         1          2  
4         0         0         1          1  
5         0         0         1          1  
6         1         0         1          2  
7         1         1         0          2  
8         0         1         0          1  
9         0         0         1          1

In [3]:
df.shape

(59776, 17)

In [97]:
df['class'].unique()

array(['Necromancer', 'Slayer', 'Deadeye', 'Gladiator', 'Inquisitor',
       'Raider', 'Champion', 'Occultist', 'Pathfinder', 'Elementalist',
       'Chieftain', 'Hierophant', 'Ascendant', 'Trickster', 'Guardian',
       'Berserker', 'Juggernaut', 'Saboteur', 'Assassin', 'Witch',
       'Marauder', 'Ranger', 'Scion', 'Duelist', 'Shadow', 'Templar'],
      dtype=object)

In [102]:
nec30 = df[(df['class'] == 'Necromancer')&(df['rank']<=30)][['rank', 'class', 'id', 'level', 'account', 'challenges', 'ladder', 'twitch']]
nec30

rank        class                                                 id  \
0       1  Necromancer  3dcddd59f5088893f734f39686350990dae168cc4f4b00...   
1       1  Necromancer  8f3216db5ac9106c287a834731aafc83c387138f28fc45...   
2       1  Necromancer  c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...   
8       3  Necromancer  00c0030fc548cc4c0d804564f8fb356610044b7e6d484f...   
19      5  Necromancer  bace7ed50cbc71a3823c0728c8199d3af37e39042b7cbe...   
20      6  Necromancer  4b3e32dae13d4e79ac18fa6b40054d89c217f51ffecd39...   
23      6  Necromancer  ded09943fa1334fd06626316cd82fc2c6d7722dffbc2b0...   
24      7  Necromancer  221ff99fc8976f803cb8316cd9f828b208e07dacabf4a1...   
31      8  Necromancer  5f33abb3669a3fee580d42fa54333f51b3a9ee390a41df...   
36     10  Necromancer  14ae61654818240433967edfb4fe3cb9732224af9d4b31...   
38     10  Necromancer  919500f447943e0e85df5ff8fbbe37cfc63ab8eb76c116...   
42     11  Necromancer  c8d8d1184c7a16c3ce6e5c93b4c3c441ab91690b5b2a9f...   
45     12  Necromancer  6a2bc2f33c4b6a549a6a9293ba854b6af5e4c995b07a4d...   
47     12  Necromancer  ed05d6028113adea628ea09bce3530ee9636c4a0502808...   
50     13  Necromancer  c99ee312bf4e3bcff349542e455e1c4e9979e0ce8afaed...   
52     14  Necromancer  3bebf29a9206a1d1d0444800cecfed40ee83ec65281279...   
53     14  Necromancer  44d24cd3203b41d6c2ca5d9c40c6be78d82ceb12c66fe3...   
56     15  Necromancer  01570a32397beab3e611b592e02c5a9761b81c37384a96...   
62     16  Necromancer  455c61c7f00ab39a754e81c85a29a646ac132085c09f02...   
77     20  Necromancer  7d1d0a24acd2f54ee2a4160484cba937ea906a580a3c7b...   
79     20  Necromancer  aeea7345a66b432bb05b4621d29048196bdcd3a77b35d4...   
87     22  Necromancer  96c8ff98bac924e91673f5745036d94676e03479c46ba8...   
90     23  Necromancer  f6f024200acfa75990cea5778bcc13ff12a52f4e19a8f5...   
97     25  Necromancer  3eea999461293c873f1a3fa502f5919afc4dbc23bc11c9...   
104    27  Necromancer  39e68e5ba9c48ab01d7130309c9acfab88c5e9cfe515e8...   
108    28  Necromancer  2e7bd8ebf95ec750ee9b1a71c43b158e9dee7e35e8ca49...   
112    29  Necromancer  0b8fde607f3aaf2ba0e2265580f1fb4521f0d6061260a8...   
117    30  Necromancer  8961a59c3e206de93ccb5f4e8c9bf927948fc620ec4626...   
118    30  Necromancer  bccee2f9ec1cc967ebf20a2a1e841f4391d0296e72ab57...   

     level           account  challenges              ladder            twitch  
0      100            TheTzn          40           Harbinger            thetzn  
1      100            RaizQT          22    SSF Harbinger HC            raizqt  
2      100       GucciPradas          35  Hardcore Harbinger       guccipradas  
8      100          rami1337          40           Harbinger          rami1337  
19     100           Pochtli          40       SSF Harbinger               NaN  
20     100            morale          40           Harbinger         akamorale  
23     100        Toaster404          36  Hardcore Harbinger  shockingtoasters  
24     100            ashmed          29    SSF Harbinger HC               NaN  
31     100        Laticauda1          36       SSF Harbinger        laticaudaz  
36     100           bannyan          28    SSF Harbinger HC          barnyang  
38     100           SkiTTyy          27  Hardcore Harbinger           skitty0  
42     100          chistor_          40  Hardcore Harbinger          chistor_  
45     100          chefstef          40       SSF Harbinger      chefstef_y2k  
47     100            Realyn          38    SSF Harbinger HC           nyylaer  
50     100            Skyevn          34           Harbinger        tranhgiang  
52     100    pathofexilegod          33           Harbinger    epictelevision  
53     100         jonas1234          25    SSF Harbinger HC               NaN  
56     100            nikoon          28    SSF Harbinger HC               NaN  
62     100       xxownxagexx          24  Hardcore Harbinger               NaN  
77      99           Trakand          36    SSF Harbinger HC           tra

In [104]:
nec30.describe()

rank       level  challenges
count  29.000000   29.000000   29.000000
mean   14.448276   99.517241   32.758621
std     9.279725    0.828971    5.895853
min     1.000000   98.000000   22.000000
25%     7.000000   99.000000   28.000000
50%    13.000000  100.000000   34.000000
75%    22.000000  100.000000   38.000000
max    30.000000  100.000000   40.000000

In [105]:
# 30위 안에 드는 네크로멘서들이 ladder division 별로 몇 명이나 있냐. 

nec30.groupby(by = 'ladder').nunique()

rank  class  id  level  account  challenges  twitch
ladder                                                                 
Harbinger              6      1   6      1        6           4       6
Hardcore Harbinger     6      1   6      1        6           6       5
SSF Harbinger          6      1   6      3        6           5       2
SSF Harbinger HC      11      1  11      3       11           9       7

근데 이들, 생각보다 랭크가 높다. 무려 세개의 부문에서 1등을 하고 있음.

30등 안에 들은 비율들을 클래스별로 내본다.

그다음 총 클래스가 26개인 것을 고려해서 최소 ladder 별로 1퍼센트 이상이 되면 꽤 많은 것 이런식으로 가도 될 것 같고.

30등 안에 들을 확률이 높은 일부 랭크 안에서도 구해서 가중? 치를 두어도 될 것 같다는 생각이 들음.

# Let's unique user analyzing!
---
주로 이용할 컬럼은 account, id, name, twitch이다.

이전에 데이터셋을 탐색하며 보았듯, account는 유저의 계정 자체를 의미하며, id는 캐릭터 자체 고유아이디. name은 해당 캐릭터의 이름이며, twitch는 유저가 선택적으로 입력하는 자신의 twitch아이디이다. twitch 컬럼은 유일하게 Nan 값이 있었던 컬럼이기도 하다.

In [4]:
df[['account', 'name', 'id', 'twitch']].describe()

account    name                                                 id  \
count     59776   59776                                              59776   
unique    37123   58286                                              59776   
top     chuz101  _Flic_  3dcddd59f5088893f734f39686350990dae168cc4f4b00...   
freq         30      14                                                  1   

                  twitch  
count              16300  
unique              9613  
top     thespunkylobster  
freq                  18

In [5]:
# 전체 고유 값 개수

df.nunique()

rank             14981
dead                 2
online               2
name             58286
level               48
class               26
id               59776
experience       54124
account          37123
challenges          37
twitch            9613
ladder               4
special_class        2
stat_Str             2
stat_Int             2
stat_Dex             2
stats_num            3
dtype: int64

In [6]:
# 결측 확인

df.isnull().sum()

rank                 0
dead                 0
online               0
name                 0
level                0
class                0
id                   0
experience           0
account              0
challenges           0
twitch           43476
ladder               0
special_class        0
stat_Str             0
stat_Int             0
stat_Dex             0
stats_num            0
dtype: int64

In [7]:
# 고유 값 개수 정리

print('Account: ', df['account'].nunique())
print('Name: ', df['name'].nunique())
print('Id: ', df['id'].nunique())
print('Twitch: ', df['twitch'].nunique())

Account:  37123
Name:  58286
Id:  59776
Twitch:  9613


# About 'name'
---

id는 고유하지만 name은 겹치는 것들이 분명 있었다. 캐릭터 name을 정하는데에 있어서 제한이 있는 것으로는 보이지 않는다.

간단히 살펴본다.

In [8]:
# 고유값 확인

df['name'].nunique()

58286

In [9]:
df['name'].sample(5)

41189    SluttySexFestivalBTW
8912                  EleProj
14131          Рабочий_Пончик
11345              innotavain
58761         SHIZUMETEYARUKA
Name: name, dtype: object

In [65]:
df['name'].value_counts().sort_values(ascending=False)

_Flic_            14
TundraDrum        11
LordTemplar       10
SirLalli           9
kotlecio           9
                  ..
ОгоньОчищения      1
AzertyTheWitch     1
pfooey             1
Politoese          1
Reselin            1
Name: name, Length: 58286, dtype: int64

In [84]:
df[['account','name', 'id', 'twitch']].groupby(by = 'name').count().sort_values('account',ascending=False)

account  id  twitch
name                                     
_Flic_                     14  14       0
TundraDrum                 11  11       0
LordTemplar                10  10       0
kotlecio                    9   9       0
SirLalli                    9   9       0
...                       ...  ..     ...
Ikasa                       1   1       0
IkenathSSFwithNEWISP        1   1       0
IkenathStockpiled           1   1       0
IkezorFaceRekt              1   1       0
๛๑bOrbEnz๑๛                 1   1       0

[58286 rows x 3 columns]

이름 별로 다른 account를 가진, 말그대로 혹시나 다른 사람일 수 있는 계정들을 불러와보고 싶었는데 마음대로 되지 않는다....


자연어처리 능력이 뛰어나면 같이 닉네임을 맞춘 사람들도 구할 수 있을 거라는 생각도 든다.

##### 가장 확실한 실마리는  'account'이다.

# Account tell us...
---

In [10]:
df['account'].value_counts().sort_values(ascending=False)

chuz101       30
HQuality      29
TundraDrum    25
Hyperized     25
Ardwage       23
              ..
Qoom           1
Migiri         1
Shakaja        1
Clocksimus     1
woworpoe       1
Name: account, Length: 37123, dtype: int64

account는 총 37123개로, 랭크에 올라있는 약 6만여개의 캐릭터들 중 한 유저가 여러 캐릭터를 랭킹에 올린 것을 정황상 알 수 있는 것이다.

## same account, different id

한 유저가 만든 여러 id들에 대해 살펴보자.

In [11]:
df[['account','name', 'id', 'twitch']].groupby(by = 'account').count()

name  id  twitch
account                             
..A_A                  6   6       0
0010                   2   2       0
001010011100101110     1   1       1
007JBond               1   1       1
00fant                 7   7       0
...                  ...  ..     ...
鹿島アントラーズFC_強化部_鈴木満     1   1       0
黒い太陽                   1   1       1
김태연                    3   3       3
누여김                    3   3       0
람색스                    1   1       0

[37123 rows x 3 columns]

In [12]:
df[['account', 'name', 'id']].groupby(['account'])['id'].count().sort_values(ascending=False).head(30)

account
chuz101             30
HQuality            29
TundraDrum          25
Hyperized           25
Ardwage             23
He_cTou_3a_MHou     20
l3ifson             20
CosmicWonder        19
Glenoidalis         19
shaar               19
flexxorzz           19
Fuzzballs           18
Elimoku             18
Develina            18
Maha_Bone           18
swayolo             18
inshole             18
flanelostopy        18
TheSpunkyLobster    18
yekarnybabaj        17
Mousseman           17
drgoatphd           17
Hellrazah           16
Seyser_Koze         16
raftner             16
pronemad            16
beuyapode           16
Ezhiel              16
Tunechis            15
turboteddy          15
Name: id, dtype: int64

한 유저가 최대 30개의 캐릭터를 랭킹에 올려놓은 것으로 보인다. 
### account = 'chuz101'
chuz101이라는 유저를 보자.

In [93]:
# account = 'chuz101' are 30.

df[df['account'] == 'chuz101'].head(10)

rank   dead  online                    name  level         class  \
10045  2516   True   False             Chuz_Sunder     84        Slayer   
10242  2565   True   False             Chuz_WChief     84     Berserker   
11167  2796   True   False  Chuz_VisLinkedByChance     83   Necromancer   
12471  3123   True   False           Chuz_OroChief     82     Chieftain   
13523  3387   True   False               Chuz_Glad     81     Gladiator   
13853  3470   True   False               Chuz_Farm     81        Slayer   
14028  3514  False   False                 Chuz_EA     81  Elementalist   
14110  3534   True   False           Chuz_FBTotems     81    Hierophant   
15508  3885   True   False                Chuz_SRS     80   Necromancer   
19139  4795   True   False             Chuz_Zerker     77     Berserker   

                                                      id  experience  account  \
10045  9e49cb7823b454812b066f3c1874efdbac477c8d18d17b...  1232028776  chuz101   
10242  93c1f129fbbdf2fda65d42fbd13c75fa39b2f6445c836c...  1216709757  chuz101   
11167  fa59e96e7142c23f8c531b085c73107f38f879e8322578...  1146254611  chuz101   
12471  fc8eddd1529977eefa4af1bf8f2f70c7e6037f8b1cc4cc...  1050328461  chuz101   
13523  1c96b4fcdef28c42219bd0f6663ff7efd89b779b75ba2b...   979602294  chuz101   
13853  567d7754d24d403983d29d75c08f67960d959178215eb8...   958655183  chuz101   
14028  06e11c02e8d8c118fdd91aa0135540d5655806da3fa370...   948202435  chuz101   
14110  b19dde189743f38138e5aa0dff1c4ff315af6c82e4a3a3...   943703466  chuz101   
15508  3d6234041d9d7c843932870441a89439918aea7521ee75...   865544505  chuz101   
19139  2c9b1740c2ebe9a24bc894c0ac28081535ca3b1038ee7b...   686077848  chuz101   

       challenges twitch            ladder  special_class  stat_Str  stat_Int  \
10045          25    NaN  SSF Harbinger HC              1         1         0   
10242          25    NaN  SSF Harbinger HC              1         1         0   
11167          25    NaN  SSF Harbinger HC              1         0         1   
12471          25    NaN  SSF Harbinger HC              1         1         0   
13523          25    NaN  SSF Harbinger HC              1         1         0   
13853          25    NaN  SSF Harbinger HC              1         1         0   
14028          25    NaN  SSF Harbinger HC              1         0         1   
14110          25    NaN  SSF Harbinger HC              1         1         1   
15508          25    NaN  SSF Harbinger HC              1         0         1   
19139          25    NaN  SSF Harbinger HC              1         1         0   

       stat_Dex  stats_num  
10045         1          2  
10242         0          1  
11167         0          1  
12471         0          1  
13523         1          2  
13853         1          2  
14028         0          1  
14110         0          2  
15508         0          1  
19139         0          1

In [14]:
df[df['account'] == 'chuz101'].nunique()

rank             30
dead              2
online            1
name             26
level            19
class            11
id               30
experience       30
account           1
challenges        1
twitch            0
ladder            2
special_class     1
stat_Str          2
stat_Int          2
stat_Dex          2
stats_num         3
dtype: int64

In [15]:
# 어떤 이름들을 사용하고 있나?

df[df['account'] == 'chuz101'][['name', 'twitch', 'class', 'ladder']]

name twitch         class              ladder
10045             Chuz_Sunder    NaN        Slayer    SSF Harbinger HC
10242             Chuz_WChief    NaN     Berserker    SSF Harbinger HC
11167  Chuz_VisLinkedByChance    NaN   Necromancer    SSF Harbinger HC
12471           Chuz_OroChief    NaN     Chieftain    SSF Harbinger HC
13523               Chuz_Glad    NaN     Gladiator    SSF Harbinger HC
13853               Chuz_Farm    NaN        Slayer    SSF Harbinger HC
14028                 Chuz_EA    NaN  Elementalist    SSF Harbinger HC
14110           Chuz_FBTotems    NaN    Hierophant    SSF Harbinger HC
15508                Chuz_SRS    NaN   Necromancer    SSF Harbinger HC
19139             Chuz_Zerker    NaN     Berserker    SSF Harbinger HC
21551          Chuz_Berzerker    NaN     Berserker    SSF Harbinger HC
22296               Chuz_Fire    NaN     Berserker    SSF Harbinger HC
22797       Chuz_SRS_Take_Two    NaN   Necromancer    SSF Harbinger HC
23630              Chuz_SGlad    NaN     Gladiator    SSF Harbinger HC
24054                Chuz_SRS    NaN   Necromancer    SSF Harbinger HC
24975               Chuz_Mine    NaN      Saboteur    SSF Harbinger HC
27523              Chuz_Miner    NaN      Saboteur    SSF Harbinger HC
28253           Chuz_Ballista    NaN       Deadeye    SSF Harbinger HC
30254                Chuz_Ray    NaN    Inquisitor    SSF Harbinger HC
31880             Chuz_Sunder    NaN        Slayer    SSF Harbinger HC
32105              Chuz_Balls    NaN     Ascendant    SSF Harbinger HC
32406             Chuz_Totems    NaN    Inquisitor    SSF Harbinger HC
32613             Chuz_Totems    NaN    Inquisitor    SSF Harbinger HC
46819              Chuz_Baron    NaN   Necromancer  Hardcore Harbinger
46876            Chuz_Brotems    NaN     Berserker    SSF Harbinger HC
49537            Chuz_Spectre    NaN   Necromancer    SSF Harbinger HC
51974          Chuz_Firestorm    NaN  Elementalist    SSF Harbinger HC
53420          Chuz_BV_Zerker    NaN     Berserker    SSF Harbinger HC
58680            Chuz_Cyclone    NaN     Gladiator    SSF Harbinger HC
59703               Chuz_Glad    NaN     Gladiator    SSF Harbinger HC

In [91]:
df[df['account'] == 'chuz101']['name']

10045               Chuz_Sunder
10242               Chuz_WChief
11167    Chuz_VisLinkedByChance
12471             Chuz_OroChief
13523                 Chuz_Glad
13853                 Chuz_Farm
14028                   Chuz_EA
14110             Chuz_FBTotems
15508                  Chuz_SRS
19139               Chuz_Zerker
21551            Chuz_Berzerker
22296                 Chuz_Fire
22797         Chuz_SRS_Take_Two
23630                Chuz_SGlad
24054                  Chuz_SRS
24975                 Chuz_Mine
27523                Chuz_Miner
28253             Chuz_Ballista
30254                  Chuz_Ray
31880               Chuz_Sunder
32105                Chuz_Balls
32406               Chuz_Totems
32613               Chuz_Totems
46819                Chuz_Baron
46876              Chuz_Brotems
49537              Chuz_Spectre
51974            Chuz_Firestorm
53420            Chuz_BV_Zerker
58680              Chuz_Cyclone
59703                 Chuz_Glad
Name: name, dtype: object

In [16]:
# 이 사람이 선호하는 스탯은?

df[df['account'] == 'chuz101'][['stat_Str', 'stat_Dex', 'stat_Int']].sum()

stat_Str    19
stat_Dex    11
stat_Int    15
dtype: int64

#### Commentary

이 유저는 ladder가 둘이다. **두 가지 모드에 집중하여 플레이**를 한 것으로 예상된다. 그렇지만 실제로 뜯어보니 Necromancer를 플레이 했던 HC Harbinger 모드 외에는 전부 SSF Harbinger HC였다.

class는 총 11로, 19개+7개 총 26가지의 직업중 **11가지 클래스**를 플레이 한 것으로 나타난다.

이 사람은 30 개의 id 중 19개의 id에서 힘을 사용하는 클래스를 했다. 전직후 19개의 직업이라고 치면 힘을 선호하는 경향이 상당한 것으로 보인다.

name를 보니 자신의 유니크한 아이디 'chuz'를 앞에 붙여서 플레이 한 것으로 보인다. 

이런식으로 몇 유저를 들여다 볼 수 있을 것으로 생각된다.

### account = 'HQuality'
이번에는 29개를 가진 HQuality 라는 사람을 보자

In [17]:
# account = 'HQuality' are 29.

df[df['account'] == 'HQuality'].head(10)

rank  dead  online                     name  level        class  \
6096   1527  True   False    twmYouSeeRipHighlight     87   Juggernaut   
8963   2245  True   False             ZeroIntTrash     85    Gladiator   
12261  3071  True   False  PeaceOfMind____________     82       Slayer   
16637  4168  True   False                Anythiing     79  Necromancer   
19863  4975  True   False   BestClassBestSkillSoEz     77    Trickster   
19908  4986  True   False             ForForhayley     76   Pathfinder   
20368  5102  True   False      Z__A__E__B__A__L__O     76     Champion   
24769  6204  True   False  MinusHundreDegreeKnives     73   Inquisitor   
27369  6856  True   False  HQ_____________________     72    Gladiator   
27476  6883  True   False               GGGlovesDP     72    Berserker   

                                                      id  experience  \
6096   b8303c0d6ae52aa7b169a2ecbbf6dd25fac4797b1c989e...  1568248284   
8963   c2717e629983dbd35cae8122d8b371ee7f542aeff62a9e...  1310913363   
12261  756040d6e835bb369e3de511cf465ff0d2e9b1e206a5ed...  1068881948   
16637  a0b2981fb133633fdbd03d1bef26be10d82573aa55ed65...   808682223   
19863  c59eb000f9aced6d8bae7d7eee189ce9888bbdec6bde58...   664824416   
19908  da727f8d5c924226220255b491c46e636a607d338a06cd...   659018752   
20368  0f7e061ba240f64ae7fe5decf9122f4d35e7d3d83c1414...   638948074   
24769  60ec5c8fbcd9c968f1f54871cc8a84596062129c2c850c...   509969427   
27369  7289abe3e0325b74015f7c2c36877af726217f31131477...   452037919   
27476  1f6e37bb50ca1a1547d29fbfa9ff4ee8b78d1295f1af73...   450521584   

        account  challenges twitch            ladder  special_class  stat_Str  \
6096   HQuality          22    NaN  SSF Harbinger HC              1         1   
8963   HQuality          22    NaN  SSF Harbinger HC              1         1   
12261  HQuality          22    NaN  SSF Harbinger HC              1         1   
16637  HQuality          22    NaN  SSF Harbinger HC              1         0   
19863  HQuality          22    NaN  SSF Harbinger HC              1         0   
19908  HQuality          22    NaN  SSF Harbinger HC              1         0   
20368  HQuality          22    NaN  SSF Harbinger HC              1         1   
24769  HQuality          22    NaN  SSF Harbinger HC              1         1   
27369  HQuality          22    NaN  SSF Harbinger HC              1         1   
27476  HQuality          22    NaN  SSF Harbinger HC              1         1   

       stat_Int  stat_Dex  stats_num  
6096          0         0          1  
8963          0         1          2  
12261         0         1          2  
16637         1         0          1  
19863         1         1          2  
19908         0         1          1  
20368         0         1          2  
24769         1         0          2  
27369         0         1          2  
27476         0         0          1

In [18]:
df[df['account'] == 'HQuality'].nunique()

rank             29
dead              1
online            1
name             29
level            20
class            13
id               29
experience       29
account           1
challenges        1
twitch            0
ladder            1
special_class     1
stat_Str          2
stat_Int          2
stat_Dex          2
stats_num         3
dtype: int64

In [19]:
# 어떤 이름들을 사용하고 있나? 

df[df['account'] == 'HQuality'][['name', 'twitch', 'class', 'ladder']]

name twitch        class            ladder
6096     twmYouSeeRipHighlight    NaN   Juggernaut  SSF Harbinger HC
8963              ZeroIntTrash    NaN    Gladiator  SSF Harbinger HC
12261  PeaceOfMind____________    NaN       Slayer  SSF Harbinger HC
16637                Anythiing    NaN  Necromancer  SSF Harbinger HC
19863   BestClassBestSkillSoEz    NaN    Trickster  SSF Harbinger HC
19908             ForForhayley    NaN   Pathfinder  SSF Harbinger HC
20368      Z__A__E__B__A__L__O    NaN     Champion  SSF Harbinger HC
24769  MinusHundreDegreeKnives    NaN   Inquisitor  SSF Harbinger HC
27369  HQ_____________________    NaN    Gladiator  SSF Harbinger HC
27476               GGGlovesDP    NaN    Berserker  SSF Harbinger HC
27679     CHRIS_CYKA_GIVE_QOTF    NaN   Pathfinder  SSF Harbinger HC
29410   ImpetuousWanderOfHaste    NaN    Ascendant  SSF Harbinger HC
30351   HQ____________________    NaN   Inquisitor  SSF Harbinger HC
30752                 surprism    NaN       Raider  SSF Harbinger HC
31387             PeaceOfMind_    NaN    Occultist  SSF Harbinger HC
33099        OneGirlTwoSiphons    NaN       Raider  SSF Harbinger HC
34705  GLACIALHAMMERGREATAGAIN    NaN    Ascendant  SSF Harbinger HC
36675  R__F_________J__U__G__G    NaN   Juggernaut  SSF Harbinger HC
37190   HomoSRSectus_dibilicus    NaN   Inquisitor  SSF Harbinger HC
37675  CONTENTDESTROYERCHIEFTY    NaN    Chieftain  SSF Harbinger HC
39777                StillOpie    NaN  Necromancer  SSF Harbinger HC
40015           SURPRISEfullMF    NaN    Chieftain  SSF Harbinger HC
42025   summonragingschoolboys    NaN  Necromancer  SSF Harbinger HC
45262  stillsaneexilexilexilei    NaN       Raider  SSF Harbinger HC
47017              EK_ONE_LOVE    NaN   Inquisitor  SSF Harbinger HC
48737         ooohDARKPACToooh    NaN    Berserker  SSF Harbinger HC
51110              inHQuisitor    NaN   Inquisitor  SSF Harbinger HC
55162        WanderNumberThree    NaN       Raider  SSF Harbinger HC
56150             GGGWhySoEasy    NaN   Juggernaut  SSF Harbinger HC

In [94]:
df[df['account'] == 'HQuality'][['name', 'twitch', 'class', 'ladder']]['name']

6096       twmYouSeeRipHighlight
8963                ZeroIntTrash
12261    PeaceOfMind____________
16637                  Anythiing
19863     BestClassBestSkillSoEz
19908               ForForhayley
20368        Z__A__E__B__A__L__O
24769    MinusHundreDegreeKnives
27369    HQ_____________________
27476                 GGGlovesDP
27679       CHRIS_CYKA_GIVE_QOTF
29410     ImpetuousWanderOfHaste
30351     HQ____________________
30752                   surprism
31387               PeaceOfMind_
33099          OneGirlTwoSiphons
34705    GLACIALHAMMERGREATAGAIN
36675    R__F_________J__U__G__G
37190     HomoSRSectus_dibilicus
37675    CONTENTDESTROYERCHIEFTY
39777                  StillOpie
40015             SURPRISEfullMF
42025     summonragingschoolboys
45262    stillsaneexilexilexilei
47017                EK_ONE_LOVE
48737           ooohDARKPACToooh
51110                inHQuisitor
55162          WanderNumberThree
56150               GGGWhySoEasy
Name: name, dtype: object

In [20]:
# 이 사람이 선호하는 스탯은?

df[df['account'] == 'HQuality'][['stat_Str', 'stat_Dex', 'stat_Int']].sum()

stat_Str    18
stat_Dex    13
stat_Int    12
dtype: int64

#### Commentary 

이 유저는 name에서 유사성을 찾기 어려운 유저이다. 이 유저를 보고서, 이름간의 유사도를 가지고 다계정인지 추측해보기를 포기했다. 

이 유저는 29개의 id로 SSF Harbinger HC 모드만 한 것으로 보인다.

보면 13개의 class를 했는데, 전체적으로 훑어보면 힘과 관련된 직업을 좀 더 선호한 것으로 보인다.

실제로 각 능력치에 1씩 더하는 것을 보기만 해도 13개의 직업을 했으나 29개의 id에서 최소 18개의 직업은 힘은 찍은 것으로 나타났다.

여기서 궁금한 점은 SSF Harbinger HC 를 플레이한 유니크 account가 궁금해진다. (물론 각 모드별 unique account 비율이 궁금하다.)

## unique account in each 'ladder'

In [21]:
# unique id

df.pivot_table(index = 'ladder', values = 'id', aggfunc = 'nunique')

id
ladder                   
Harbinger           14918
Hardcore Harbinger  14905
SSF Harbinger       14981
SSF Harbinger HC    14972

In [22]:
# unique account

df.pivot_table(index = 'ladder', values = 'account', aggfunc = 'nunique')

account
ladder                     
Harbinger             13160
Hardcore Harbinger     9020
SSF Harbinger         10214
SSF Harbinger HC       6374

In [106]:
# total unique user rate 

df['account'].nunique() / df['id'].nunique()

0.6210351980728052

In [107]:
# list for each ladder unique user rate 

ladders = df['ladder'].unique()
ladders.tolist()

['Harbinger', 'SSF Harbinger HC', 'Hardcore Harbinger', 'SSF Harbinger']

In [115]:
# ladder 별 유니크 유저 비율

for i in ladders:
    v = df[df['ladder'] == i]['account'].nunique() / df[df['ladder'] == i]['id'].nunique() * 100
    print(' ladder: {} \n > unique users rate: {}'.format(i, f'{v:.2f}%'))

 ladder: Harbinger 
 > unique users rate: 88.22%
 ladder: SSF Harbinger HC 
 > unique users rate: 42.57%
 ladder: Hardcore Harbinger 
 > unique users rate: 60.52%
 ladder: SSF Harbinger 
 > unique users rate: 68.18%


### Commentary 

여러 캐릭터를 보유하고 있는 두 account를 살펴본 결과, ladder가 SSF Harbinger HC에 집중되어있는 것을 보았다. 

그로서 유추할 수 있는 것은 'SSF Harbinger HC'에서는 one account - many id 비율이 높을 것이라 생각했는데 실제였다.

높은 난이도를 요구할 것으로 추측되었던 SSF Harbinger HC 모드에서는 실제로 unique한 account 비율이 상당히 낮았다.

일단, 랭킹 자체가 id(캐릭터) 자체로 올라가는 것이다보니 이런 일이 가능한 것이고,

잘하는 유저들의 실력과 그외의 유저들과의 실력이 격차가 있는 것일 수도 있다. 

일반모드는 유니크 유저 비율이 88%에 이르지만 다른 모드들은 다르다.

랭커들이 다른 캐릭터들(id)을 더 만들어서 더 도전하게 하는 이유가 따로 있을지도 모른다(랭커들에게만 주어지는 보상같은 것).

## same name, different id

한 가지 이름으로 유저는 어떻게 알아볼 수 있을 까

In [26]:
df[['account','name', 'id', 'twitch']].groupby(by = 'name').count()

account  id  twitch
name                                        
AAAAAAAAAAAAAMUTALA            1   1       0
AAAAAAAAAAAAyourself           1   1       1
AAAAAAAAAAHHHHH                1   1       0
AAAAAAAAASUNDERAAAAAAAA        1   1       1
AAAAAAAAAlIENNNNNNNNN          1   1       0
...                          ...  ..     ...
๒๏๏๒                           1   1       1
๓๏๓                            2   2       2
๖Neferpitou๙                   1   1       0
๛คูJชาe๛ฮี                     1   1       1
๛๑bOrbEnz๑๛                    1   1       0

[58286 rows x 3 columns]

In [27]:
df[['account', 'name', 'id', 'twitch']].groupby(['name'])['id'].count().sort_values(ascending=False).head(30)

name
_Flic_             14
TundraDrum         11
LordTemplar        10
kotlecio            9
SirLalli            9
dopenesh            8
Hellrazah           8
Znoz                8
Cya                 8
Ekainen             7
Flashow             7
HarbinGer_Meng      7
Lodystek            7
HarbinGer_Mango     7
xWolhaiksong        7
This                6
Ryacchi             6
nextinthelane       6
Xiberia             6
Kokkuwi             6
uranari             6
Khob                6
Vorkath             6
VACK_HHC            6
Pingeee             6
Elki                6
Cursecoin           6
gMax                6
OvarianPillager     6
Mimic_SSF           6
Name: id, dtype: int64

같은 이름을 최대 14개의 캐릭터가 사용중이다. 이것은(_ Flic _) 한 account일까?

In [28]:
# name = '_Flic_' are 14.

df[df['name'] == '_Flic_'].head(10)

rank  dead  online    name  level      class  \
27206   6815  True   False  _Flic_     72  Ascendant   
28071   7032  True   False  _Flic_     72  Ascendant   
28913   7243  True   False  _Flic_     71  Ascendant   
36978   9263  True   False  _Flic_     68  Ascendant   
40020  10024  True   False  _Flic_     66  Ascendant   
41317  10349  True   False  _Flic_     65  Ascendant   
42076  10539  True   False  _Flic_     65  Ascendant   
42405  10622  True   False  _Flic_     65  Ascendant   
44372  11116  True   False  _Flic_     63  Ascendant   
46567  11665  True   False  _Flic_     61  Ascendant   

                                                      id  experience  \
27206  67bcd417afe0b89cfc308b26bf590b511180130f680583...   455996734   
28071  7ec0dc750d81ebacf2ca2cce9879da1d801f0631444d25...   440263744   
28913  9bd5774ed0f816b13e97673498d13aa82cbda6ebcea9c2...   424709581   
36978  72e45259ccd8f39f0582fe90c4f518bcd612b38d591bd8...   314784438   
40020  4f1ccb10064cfb440ac865c3c2aaa658ecbcd581374b3e...   273456954   
41317  11b3c4652b8198d542f3da11b08d7775b3460937a188b6...   253743465   
42076  e946cd6ebc3e957f39860a7d7f39efc0b632512d5b4fe5...   240160428   
42405  0633f8211bb7fe74aa2fba4316ee68f76db8f013af796c...   233709956   
44372  0003c4ccac35b1f8c69a8b5b18701bf1a7310800ac1f56...   196299619   
46567  3eda4821da06924fe80ef166ff5f00fd639d4d0c611cf4...   166472598   

               account  challenges twitch            ladder  special_class  \
27206  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
28071  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
28913  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
36978  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
40020  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
41317  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
42076  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
42405  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
44372  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   
46567  He_cTou_3a_MHou          20    NaN  SSF Harbinger HC              1   

       stat_Str  stat_Int  stat_Dex  stats_num  
27206         1         1         1          3  
28071         1         1         1          3  
28913         1         1         1          3  
36978         1         1         1          3  
40020         1         1         1          3  
41317         1         1         1          3  
42076         1         1         1          3  
42405         1         1         1          3  
44372         1         1         1          3  
46567         1         1         1          3

In [29]:
df[df['name'] == '_Flic_'].nunique()

rank             14
dead              1
online            1
name              1
level            10
class             1
id               14
experience       14
account           1
challenges        1
twitch            0
ladder            1
special_class     1
stat_Str          1
stat_Int          1
stat_Dex          1
stats_num         1
dtype: int64

### Commentary

이 Flic이라는 아이디를 쓰는 유저는 account 고유 값이 1이므로 한 사람이다. twitch 아이디는 적혀있지 않지만 account가 한개인 것을 알 수 있다. 

우연으로 겹친 아이디도 분명 있을 수도 있을것이라 생각하지만 그렇지 않았다.

또한 이 사람은 아까 account == 'chuz101'인 사람과 달리, **한 모드에서 한 가지 클래스만** 집중적으로 한 것으로 보인다. 



## same twitch, different id

이번엔 같은 twitch아이디를 쓰는데 여러 캐릭터를 운영중인 유저를 찾아볼 예정이다.

어쩌면 다계정을 운영중인지에대한 여부도 알 수 있을 것이라 생각된다. (두근두근)

In [117]:
# 한 twitch 아이디로 여러 id를 운영중인 사람들

df['twitch'].value_counts().sort_values(ascending=False)

thespunkylobster    18
iispasov            16
seyser_koze         16
yoshixt             15
egheo17             15
                    ..
jigalopoe            1
taggarn              1
ragnion              1
kimitsugo            1
lecarn               1
Name: twitch, Length: 9613, dtype: int64

한 twitch 아이디 안에 있는 여러 이름과 아이디들을 보자.

가장 많은 수를 기록하고 있는 thespunkylobster 를 보도록 해보자.

In [118]:
# twitch = thespunkylobster

df[df['twitch'] == 'thespunkylobster'].head(10)

rank   dead  online                     name  level        class  \
487     122   True   False  ThroughThePrismOfSkooma     94       Raider   
495     124   True   False      Khajiit_Is_Innocent     94    Gladiator   
1026    257   True   False         LobstersOnSkooma     93       Slayer   
2887    722   True   False            WindsOfSkooma     90       Raider   
4965   1243  False   False       ForTheLoveOfSkomma     88       Slayer   
5612   1405   True   False        SpacedOutOnSkomma     87   Pathfinder   
8675   2173  False   False               SkoomaRage     85  Necromancer   
8744   2190   True   False  SkoomaCausesBadDecision     85      Deadeye   
9659   2419  False   False              DarthSkooma     84    Berserker   
10299  2579   True   False          SkoomaMeatballs     84       Slayer   

                                                      id  experience  \
487    0be0d070caf5d4254fb2bd731c2c2b5fead2e1db19cbcf...  2806353782   
495    07ba54024352737d8693d61794a1e21f7ff60d4a7c556f...  2798624220   
1026   05c0abe82318cd4fdc20142d73b4ab4af36f24b32c94bd...  2465132591   
2887   ce0931dea5c235a231be0487cd51aed0f82bf8737b1ce2...  1955389953   
4965   263ea5b2089de5383dd9e0eff5d308746bc1a6c362044e...  1689357993   
5612   8cb6391296e8b7916cc1174195bf261d9ac4103f7abab1...  1619835792   
8675   ddf286fe358060d8dc3eac59e61541051b774fee6ff127...  1335048935   
8744   dfa99a355be72cc79fe54b0312b623df367ed4a05f2cd5...  1328292006   
9659   db7f3565e482bdd84e0589f7c6685ad7e78273f234a9d3...  1260634487   
10299  d9fdf5925af585ef933e992b17a31585ca483bbc3b5d83...  1213016381   

                account  challenges            twitch            ladder  \
487    TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
495    TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
1026   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
2887   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
4965   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
5612   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
8675   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
8744   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
9659   TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   
10299  TheSpunkyLobster          30  thespunkylobster  SSF Harbinger HC   

       special_class  stat_Str  stat_Int  stat_Dex  stats_num  
487                1         0         0         1          1  
495                1         1         0         1          2  
1026               1         1         0         1          2  
2887               1         0         0         1          1  
4965               1         1         0         1          2  
5612               1         0         0         1          1  
8675               1         0         1         0          1  
8744               1         0         0         1          1  
9659               1         1         0         0          1  
10299              1         1         0         1          2

In [32]:
df[df['twitch'] == 'thespunkylobster'].nunique()

rank             18
dead              2
online            1
name             16
level            13
class             9
id               18
experience       18
account           1
challenges        1
twitch            1
ladder            1
special_class     1
stat_Str          2
stat_Int          2
stat_Dex          2
stats_num         2
dtype: int64

### Commentary

우선 궁금했던 account의 수는 1개이다. 다계정을 운영하지는 않는 것으로 보인다. 

이 유저의 **name 을 보면 작은 규칙**이 있다. Skooma, Skomma, Lobster(s), Khajiit 등의 표현을 넣고 있으며, 그런 표현은 한 번만 등장하지 않는다. 

이러한 점들을 보아서는 이 트위치 유저가 여러개의 캐릭터를 관리하며 랭킹에 이름을 올렸을 가능성이 있다. 

(이 트위치 아이디가 적힌 것이 해당 유저의 팬들이 따라서 쓰거나 사칭하며 랭킹에 이름을 올린게 아니라면 말이다.)

특이한 점은 이 유저는 SSF Harbinger HC 즉 **혼자 하는 Hardcore 모드**에만 등장한다는 점이다. 또한 HC모드는 dead 여부가 두 가지로 구분되는데 살아남은 상태인 것도 있는 것으로 보인다.

그런 솔로 플레이를 twitch에 방송을 하고 있었을 가능성이 있다.

또한 class는 총 9개로 18개의 캐릭터 중 겹치기도 하면서 여러 클래스를 플레이 했을 것이고, special_class컬럼이 고유가1개인데 9개 클래스인 것으로 보아 모든 캐릭터는 전직을 한 상태인 것으로 보인다.

# 한 계정에서 동시 접속이 가능한가?

In [123]:
# 한 계정안의 online 여부 개수

df[['account', 'name', 'id', 'online']]

account                    name  \
0            TheTzn      Tzn_NecroIsFineNow   
1            RaizQT        RaizNeverFirstQT   
2       GucciPradas  GucciStreamerAdvantage   
3          Chiroxun             ChiroxPrime   
4          cooltail     Cool_NecroIsFineNow   
...             ...                     ...   
59771  Romanchik002       ПроклятьеРекласта   
59772  Ishibashisan       IshibashiSummoner   
59773        ruby01                 BLively   
59774        Lecarn                vawddvaw   
59775      woworpoe                 Reselin   

                                                      id  online  
0      3dcddd59f5088893f734f39686350990dae168cc4f4b00...   False  
1      8f3216db5ac9106c287a834731aafc83c387138f28fc45...   False  
2      c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...   False  
3      c861372da792be0b22c45bf437ccd58437c52e9455e0bf...   False  
4      24ae924ceed7989ef3d3d6772612832bb467a6094352c7...   False  
...                                                  ...     ...  
59771  d33b4f6e08c10e365765f9a36a8f36d561fd1d86f10ee1...   False  
59772  5764cfa387e0a87a4bebc1a3c5017e92de8bbb06445c47...   False  
59773  9ac75ab75a47cee8a9dfb0a31912df8909720a8b20f9f1...   False  
59774  cf02dfc0c90b2df9c7ac76bbedd91e93c2a8a2ca629c15...   False  
59775  f7ffda5ca2490546344d32930693f8299930ce9451df4a...   False  

[59776 rows x 4 columns]

In [130]:
# 상태가 두 가지 이상인 캐릭터들은 1 이상이 True이다

simulconnect_account = df.groupby(['account'], observed = False)['online'].sum() > 1
simulconnect_account.sample(10)

account
Gyarados        False
Frynce          False
vernerek        False
MadrasBez       False
goneil          False
DGee            False
Alendert        False
Rasu            False
bazokalol       False
deadlyweasel    False
Name: online, dtype: bool

위의 simulconnect_account는 account 별로 online들의 값들 즉 True값 합이 1초과인 경우를 가져온다.

한 account 내 online의 True가 한개인 경우는 한 account 내에서 여러 캐릭터들 중 한 캐릭터를 정상적으로 접속한 것으로 볼 수 있다는 점을 생각하여

합이 1 초과이면 한 account에서 여러 캐릭터를 접속한 상태라고 볼수 있다.

그렇다면 누가 여러 캐릭터를 동시 접속한 상태인 것일까?

In [126]:
simulconnect_account[simulconnect_account == True]

account
Syquiq       True
konicar45    True
nibarel      True
Name: online, dtype: bool

In [127]:
simulusers= simulconnect_account[simulconnect_account == True].index
simulusers

Index(['Syquiq', 'konicar45', 'nibarel'], dtype='object', name='account')

In [128]:
# 위의 캐릭터들을 데이터 셋에서 가져온다.

for i in simulusers:
    arg = df['account'] == i
    Frame = df[arg][['account', 'name', 'id', 'online']]
    print(Frame)

      account                 name  \
3640   Syquiq       Sisyphus_Spins   
19330  Syquiq  Ashura_The_Survivor   
26370  Syquiq  Parallax_The_Purple   

                                                      id  online  
3640   5206a212ad4032da92b139ae634c9c288a4ad832f4ee74...    True  
19330  bd4625d8ed1c6a71accd3906f4e77f9d5eb86f12495a5d...   False  
26370  78a3bc9934d3684168548a8f3c3b991d45f3a1e78da22d...    True  
         account                 name  \
37270  konicar45         CuraBezDecka   
42938  konicar45         MlohavaCuna_   
47761  konicar45    KuracPalacP__Sisa   
59428  konicar45  RaspaljenaCrnogorka   

                                                      id  online  
37270  f9d826d704bc413f0df33241fc5880892ea2696e9602d8...    True  
42938  50acd83c75768b3d5436a342c49dcd9a1a333846838bb2...   False  
47761  4dfb907f0fad5b2ee6583d4c81bc21f73466ec8719d5da...   False  
59428  5ddb92de80d773140a32ef15a2db9735b129a0753f3a6c...    True  
       account                     nam

In [145]:
blacklist_users = list(simulusers)
# bin_df = pd.DataFrame()

for u in range(len(blacklist_users)):
    acc = blacklist_users[u]
    if u == 0:
        bin_df = df[df['account'] == acc]
    elif u != 0:
        bin_df2 = df[df['account'] == acc]
        whole_df = pd.concat([bin_df, bin_df2], axis = 0)
        bin_df = whole_df

In [146]:
bin_df

rank   dead  online                     name  level        class  \
3640     912  False    True           Sisyphus_Spins     92   Inquisitor   
19330   4843  False   False      Ashura_The_Survivor     87    Ascendant   
26370   6607  False    True      Parallax_The_Purple     84    Berserker   
37270   9299  False    True             CuraBezDecka     90   Pathfinder   
42938  10710   True   False             MlohavaCuna_     90    Berserker   
47761  11914   True   False        KuracPalacP__Sisa     90    Berserker   
59428  14819  False    True      RaspaljenaCrnogorka     89    Ascendant   
4008    1002  False    True  GentlemanGodInTheMaking     95     Guardian   
24549   6129  False   False          GentlemanLabGod     91   Juggernaut   
25473   6359  False    True      GentlemanNecromancy     91  Necromancer   
46079  11493  False   False           SPICYMEATABAWL     90    Ascendant   
52720  13153  False   False         GentlemanZuxilon     89   Juggernaut   
55255  13779   True   False             GentlemanRIP     89    Berserker   

                                                      id  experience  \
3640   5206a212ad4032da92b139ae634c9c288a4ad832f4ee74...  2268605368   
19330  bd4625d8ed1c6a71accd3906f4e77f9d5eb86f12495a5d...  1519130326   
26370  78a3bc9934d3684168548a8f3c3b991d45f3a1e78da22d...  1262399383   
37270  f9d826d704bc413f0df33241fc5880892ea2696e9602d8...  2021009512   
42938  50acd83c75768b3d5436a342c49dcd9a1a333846838bb2...  1958247722   
47761  4dfb907f0fad5b2ee6583d4c81bc21f73466ec8719d5da...  1934009687   
59428  5ddb92de80d773140a32ef15a2db9735b129a0753f3a6c...  1789766633   
4008   e5e90b76867bb23fa54a716fc4f15fdeb5ff253d2b351f...  3069192586   
24549  3bc4d6fe776886e7d9e2af52d6f07f8497ae6427abdc0c...  2227762325   
25473  1179e480c30e9610d489410c88fa2a8926d8673c3a2276...  2206120070   
46079  da53763bd4fce2ad12304c3365ff9a1e678b227e95752e...  1935497126   
52720  973ea3c949c6bb87e545910a3253b16f59fe8805c5d29d...  1854834877   
55255  df36c6c237ff52cf624644baac39595d754a10c544905b...  1828273570   

         account  challenges   twitch              ladder  special_class  \
3640      Syquiq          25      NaN       SSF Harbinger              1   
19330     Syquiq          25      NaN       SSF Harbinger              1   
26370     Syquiq          25      NaN       SSF Harbinger              1   
37270  konicar45          30  konicar  Hardcore Harbinger              1   
42938  konicar45          30  konicar  Hardcore Harbinger              1   
47761  konicar45          30  konicar  Hardcore Harbinger              1   
59428  konicar45          30  konicar  Hardcore Harbinger              1   
4008     nibarel          30      NaN  Hardcore Harbinger              1   
24549    nibarel          30      NaN  Hardcore Harbinger              1   
25473    nibarel          30      NaN  Hardcore Harbinger              1   
46079    nibarel          30      NaN  Hardcore Harbinger              1   
52720    nibarel          30      NaN  Hardcore Harbinger              1   
55255    nibarel          30      NaN  Hardcore Harbinger              1   

       stat_Str  stat_Int  stat_Dex  stats_num  
3640          1         1         0          2  
19330         1         1         1          3  
26370         1         0         0          1  
37270         0         0         1          1  
42938         1         0         0          1  
47761         1         0         0          1  
59428         1         1         1          3  
4008          1         1         0          2  
24549         1         0         0          1  
25473         0         1         0          1  
46079         1         1         1          3  
52720         1         0         0          1  
55255         1         0         0          1

한 유저가 동시에 접속이 가능한가 본 것인데, 실제로 한 유저 안에서 online이 두개인 캐릭터들이 존재했다. 

존재했다는 것만으로 이상함을 느끼는데, 실제로 문제가 되는 게임규정이라면 이 유저들을 상태로 검색해 볼 수 있을 것 같다.

# One person, many account?
---

한 계정에 최대 30개의 캐릭터도 운영중인 것으로 보면 여러 캐릭터를 키우는데 있어서 제한이 있거나 그로 인한 문제가 생길 수 있는 게임은 아닌 것으로 보인다.

보통 RPG의 경우 게임 내에서 특정 행위로 부정하게 이익을 얻는 행위를 막기위해 다계정을 막는 경우가 많다. (게임 내 경제를 위해 그런 경우가 많다.)

또한 랭킹이 기록되는 경우 실력은 그렇지 않지만 랭킹에 이름을 올리기 위해 '대리'랭크 게임을 사주하거나 사주받는 경우도 더러 있기 때문에 막는 경우가 많다.

POE에서 문제가 되는 것일지는 모르겠지만, 비슷한 데이터가 주어졌을떄 이를 분류해낼 수 있는 방법을 생각해보았다.

정확히 알 수는 없지만 정황상이며, 한 가지 조건만으로는 어뷰징 유저라고 칭할 수 없고, 여러 정황이 함께 겹치면 그런 것으로 볼 수도 있다는 것이다.

name이나 account, twitch는 사용자가 직접 정하는 부분이기 때문에 특정 단어나 뉘앙스에서 해당 유저의 아이덴티티가 드러나게 되어있다. 

1. 한 account내에서 쓰이던 비슷한 name이 아닌 **다른 name**
2. 한 account내에서 집중적으로 쓰이던 클래스가 아닌 **다른 소수의 class**
3. 한 account내에서 분포된 랭크와 달리 **튀는 rank**
4. 한 account내에서 **다른 twitch**
5. 같은 name 혹은 유사한 name이나 **다른 account**
6. 같은 name 혹은 유사한 name이나 **다른 twitch**
7. 한 twitch 내에서 **다른 account**


등등이다.

분석을 진행하기에 앞서서 분석에 쓰이기 어려운 일부 column들은 과감히 내려놓으려 한다.

(online, dead)


In [38]:
df1 = df.drop(columns = ['online', 'dead'])

In [39]:
df1.head()

rank                    name  level        class  \
0     1      Tzn_NecroIsFineNow    100  Necromancer   
1     1        RaizNeverFirstQT    100  Necromancer   
2     1  GucciStreamerAdvantage    100  Necromancer   
3     1             ChiroxPrime    100       Slayer   
4     2     Cool_NecroIsFineNow    100      Deadeye   

                                                  id  experience      account  \
0  3dcddd59f5088893f734f39686350990dae168cc4f4b00...  4250334444       TheTzn   
1  8f3216db5ac9106c287a834731aafc83c387138f28fc45...  4250334444       RaizQT   
2  c6ec2dae3855c551e0597c06ef2da06fbb5512487ded7f...  4250334444  GucciPradas   
3  c861372da792be0b22c45bf437ccd58437c52e9455e0bf...  4250334444     Chiroxun   
4  24ae924ceed7989ef3d3d6772612832bb467a6094352c7...  4250334444     cooltail   

   challenges       twitch              ladder  special_class  stat_Str  \
0          40       thetzn           Harbinger              1         0   
1          22       raizqt    SSF Harbinger HC              1         0   
2          35  guccipradas  Hardcore Harbinger              1         0   
3          26          NaN       SSF Harbinger              1         1   
4          22          NaN           Harbinger              1         0   

   stat_Int  stat_Dex  stats_num  
0         1         0          1  
1         1         0          1  
2         1         0          1  
3         0         1          2  
4         0         1          1

## twitch로 걸러내기

한 가지 twitch계정에서 여러 account를 운영중인 사람들이 있나?

In [40]:
df1.groupby(['twitch'])['account'].count()

twitch
001100010010       1
00mis              1
0815_noobhunter    1
0cxjwbu213c13      1
0lede4             1
                  ..
zyc0billyfr34k     1
zykenus            2
zyldragoon         1
zyrhon             1
zyrobyte           1
Name: account, Length: 9613, dtype: int64

In [41]:
manyaccout = df1.groupby(['twitch'])['account'].nunique() > 1
manyaccout.sample(10)

twitch
cherd_                  False
fredrick1337            False
amandinhapikachu        False
pingispallo_            False
deboun                  False
pipsz                   False
jorian2                 False
migge84                 False
einkrieger              False
420vapeit_vapenation    False
Name: account, dtype: bool

In [42]:
manyaccout[manyaccout == True]

twitch
aza_memes       True
easykun         True
igibek          True
jimbobkevint    True
simonthadon     True
Name: account, dtype: bool

### result

한 트위치아이디를 쓰면서 많은 account를 쓰는 사람이 5명이 걸러졌다. 

이들의 account를 보기로 하였다.

In [43]:
sametw_diffacc = manyaccout[manyaccout == True].index
sametw_diffacc

Index(['aza_memes', 'easykun', 'igibek', 'jimbobkevint', 'simonthadon'], dtype='object', name='twitch')

In [44]:
# 위의 캐릭터들을 데이터 셋에서 가져온다.

for i in sametw_diffacc:
    arg = df['twitch'] == i
    Frame = df[arg][['twitch', 'account', 'name']].sort_values('account')
    print(Frame)

          twitch    account                     name
13796  aza_memes    Aza1234  AzaSkullBeltConnoisseur
53204  aza_memes  Aza_Memes            SSFbtwXDDDDDD
        twitch         account                    name
56764  easykun  KayabaAkihikoo           NCSSF_Poddy_B
57474  easykun  KayabaAkihikoo           NCSSF_Poddy_A
38732  easykun      jahnylus14  NCSSF_EasyKunCaustic_A
       twitch account             name
7498   igibek  HurmaQ           NTKPNP
29074  igibek  HurmaQ        SosuBibas
32760  igibek  HurmaQ        СосковНет
38431  igibek  HurmaQ  CurcedPersimmon
45107  igibek  HurmaQ           NLHTTK
55884  igibek  HurmaQ           IHTMZS
367    igibek  Qwapet      Pickle_Gars
2585   igibek  Qwapet        Гуакамоля
11052  igibek  Qwapet         Qwapette
25208  igibek  Qwapet        Овсяночка
             twitch       account                     name
2820   jimbobkevint  Jimbobkevint             IFoundASword
14509  jimbobkevint  Jimbobkevint           ZoomZoomZoom__
18223  jimbobke

### Commentary


위의 twitch 아이디들은 다 계정을 사용하고 있는 twtich 유저들의 account와 캐릭터의 name들이다.

aza_memes, easykun 는 account와 name에서 각각 같은 유저일 것이라 추정되는 이름들이다.

igibek 유저는 해석하기는 어렵지만 name의 규칙성이 보이지 않았고,

jimbobkevint 유저는 RealLifePleb 라는 account 작명이 도드라지는데 따로 다른 컬럼들까지 분석이 필요해 보인다.

simonthadon 유저는 Biscos 라는 account 작명이 다른 아이디와 조금 다른 부분을 보여주는데 이또한 다른 컬럼들까지 분석이 필요해 보인다.

## account로 걸러보기

### name string splitting



In [45]:
df1[['name1', 'name2', 'name3', 'name4']] = pd.DataFrame(df1.name.str.split('_', 3).tolist())

In [46]:
df1[['name1', 'name2', 'name3', 'name4']].head(10)

name1             name2        name3 name4
0                     Tzn    NecroIsFineNow         None  None
1        RaizNeverFirstQT              None         None  None
2  GucciStreamerAdvantage              None         None  None
3             ChiroxPrime              None         None  None
4                    Cool    NecroIsFineNow         None  None
5  SteelStreamerAdvantage              None         None  None
6                    Grom  WhyGodThreeCmWhy         None  None
7                  beobio           capitan         None  None
8                    Rami               The  Negromaniac  None
9           FutonBlewAway              None         None  None

In [47]:
print(df1['name'].nunique())
print(df1['name1'].nunique())

58286
52461


In [48]:
df1['name4'].notnull().sum()

1600

각 name과 account를 자연어처리를 통해 분석을 하고 싶었지만... 꿈은 컸고 거기까지는 어려울 것 같으니 같거나 다른 것만 찾아보기로 했다.


## same name 

같은 이름인데 다른 account 를 쓰는 id가 있는지 찾아보기로 했다.

In [49]:
df1.groupby(by = 'name')['account'].nunique()

name
AAAAAAAAAAAAAMUTALA        1
AAAAAAAAAAAAyourself       1
AAAAAAAAAAHHHHH            1
AAAAAAAAASUNDERAAAAAAAA    1
AAAAAAAAAlIENNNNNNNNN      1
                          ..
๒๏๏๒                       1
๓๏๓                        1
๖Neferpitou๙               1
๛คูJชาe๛ฮี                 1
๛๑bOrbEnz๑๛                1
Name: account, Length: 58286, dtype: int64

In [50]:
samename_diffacc = df1.groupby(by = 'name')['account'].nunique() > 1

samename_diffacc.sample(10)

name
sulahrae             False
LavadoDeAutos        False
Tobeahero            False
StrayKangaroo        False
IMakeESGreatAgain    False
HongigiFPT           False
Sylph_srs            False
WanderNiko           False
zsithy               False
HCH_Akurei           False
Name: account, dtype: bool

In [51]:
samename_diffacc1 = samename_diffacc[samename_diffacc == True].index
samename_diffacc1

Index(['CantEvenSpin', 'Charekil', 'DegenDegenerate', 'KitavaPLS', 'Undabed',
       'Vorkath'],
      dtype='object', name='name')

위의 사람들은 같은 이름이지만 다른 계정인 사람들이다.

In [52]:
# 위의 캐릭터들을 데이터 셋에서 가져온다.

for i in samename_diffacc1:
    arg = df['name'] == i
    Frame = df[arg][['name', 'account', 'twitch', 'ladder', 'class', 'id', 'experience']].sort_values('account')
    print(Frame)
    print('==============================================================')

               name    account      twitch              ladder   class  \
18174  CantEvenSpin      Raxom         NaN    SSF Harbinger HC  Slayer   
38884  CantEvenSpin  jooxified  stop_qqing  Hardcore Harbinger  Slayer   

                                                      id  experience  
18174  130c7b534de5904b0f1568d2ee350e34fb0e652c644726...   732855726  
38884  e88e6f083213d7975e5a544123644448247d874da941b2...  2001125336  
           name       account twitch            ladder        class  \
48215  Charekil  Andystiles95    NaN  SSF Harbinger HC     Champion   
2858   Charekil      Charekil    NaN  SSF Harbinger HC       Slayer   
6356   Charekil      Charekil    NaN  SSF Harbinger HC  Necromancer   

                                                      id  experience  
48215  39c2acc270db5912b8da609bbc0dab07fe4e92eccc08f7...   149466067  
2858   d73a13f02c69a235e55dd176e43af8fa15e83180c6b65f...  1958939526  
6356   6421575e4c15a2dbc7fdaf16d68b7584d014790523ab17...  15422287

### Commentary

우선 name == Charekil 을 보면 같은 모드 안에 다른 클래스로 다른 id가 존재하는 것(exp가 달라서 id를 입장 같은 것으로보기는 어렵다. 물론 캐릭터를 더 키워서 직업을 바꿔 입장했을 가능성은 있다.)을 보아서 다른 직업으로 도전하는 유저들이 존재하긴 했다. 이 사람은 한 개의 계정을 더 만들었던 것이 보인다.



# 많은 계정을 가진 랭커 유저들은 비슷한 직업을 할까?
---
## by one twitch account

익숙한 직업을 할까? 그들은 특정 직업에만 능통한 편일까? 아니면 다 잘하는 편일까?

먼저 한 트위치 계정을 한 사람으로 가정하고 그들이 선택한 직업들을 보기로 한다.

In [53]:
grouped_twitch = df[['account','name', 'id', 'twitch']].groupby(by = 'twitch', as_index = 'twitch').nunique().sort_values('id', ascending = False)
grouped_twitch

account  name  id
twitch                              
thespunkylobster         1    16  18
iispasov                 1     6  16
seyser_koze              1    13  16
turbofreddy1             1    15  15
egheo17                  1    10  15
...                    ...   ...  ..
japanruinedmylife        1     1   1
japabyss211              1     1   1
janusyar                 1     1   1
janitory                 1     1   1
zyrobyte                 1     1   1

[9613 rows x 3 columns]

In [54]:
# 한 개 이상 아이디를 가진 트위치 유저들. 그 중 다 계정을 사용중인 사람들부터 쉽게 보기위해 정렬.

grouped_twitch[grouped_twitch['id']>1].sort_values('account', ascending = False, ignore_index = False)

account  name  id
twitch                            
jimbobkevint           2     9   9
aza_memes              2     2   2
simonthadon            2     3   3
easykun                2     3   3
igibek                 2    10  10
...                  ...   ...  ..
nedieth                1     3   3
nechropotence          1     2   3
noachr                 1     3   3
nawsnitchin            1     3   3
superbananab0mb        1     2   2

[3257 rows x 3 columns]

In [55]:
# 인덱싱 다시

bin_df = grouped_twitch[grouped_twitch['id']>1].sort_values('account', ascending = False).reset_index()
bin_df

twitch  account  name  id
0        jimbobkevint        2     9   9
1           aza_memes        2     2   2
2         simonthadon        2     3   3
3             easykun        2     3   3
4              igibek        2    10  10
...               ...      ...   ...  ..
3252          nedieth        1     3   3
3253    nechropotence        1     2   3
3254           noachr        1     3   3
3255      nawsnitchin        1     3   3
3256  superbananab0mb        1     2   2

[3257 rows x 4 columns]

In [56]:
# merge를 이용하여 각 유저들이 가진 직업들이나 모드 등을 데려온다

twitch_df = pd.merge(bin_df, df[['twitch', 'class', 'stat_Str', 'stat_Dex', 'stat_Int', 'ladder']])
twitch_df

twitch  account  name  id      class  stat_Str  stat_Dex  \
0        jimbobkevint        2     9   9  Gladiator         1         1   
1        jimbobkevint        2     9   9  Berserker         1         0   
2        jimbobkevint        2     9   9     Raider         0         1   
3        jimbobkevint        2     9   9     Slayer         1         1   
4        jimbobkevint        2     9   9     Raider         0         1   
...               ...      ...   ...  ..        ...       ...       ...   
9939      nawsnitchin        1     3   3  Gladiator         1         1   
9940      nawsnitchin        1     3   3  Trickster         0         1   
9941      nawsnitchin        1     3   3  Gladiator         1         1   
9942  superbananab0mb        1     2   2   Guardian         1         0   
9943  superbananab0mb        1     2   2   Guardian         1         0   

      stat_Int              ladder  
0            0    SSF Harbinger HC  
1            0           Harbinger  
2            0    SSF Harbinger HC  
3            0    SSF Harbinger HC  
4            0    SSF Harbinger HC  
...        ...                 ...  
9939         0    SSF Harbinger HC  
9940         1    SSF Harbinger HC  
9941         0    SSF Harbinger HC  
9942         1           Harbinger  
9943         1  Hardcore Harbinger  

[9944 rows x 9 columns]

In [57]:
twitch_df.groupby(by = 'twitch', as_index = False)['class'].nunique()

twitch  class
0           11thtoi      2
1      13ootyknight      2
2             14237      2
3       187demontaz      1
4     19beastmode87      1
...             ...    ...
3252        zul_dra      2
3253        zumah89      3
3254    zuxilon1337      4
3255    zvodenstein      2
3256        zykenus      2

[3257 rows x 2 columns]

In [58]:
twitch_df.groupby(by = 'twitch', as_index = False)['class'].nunique()['class'].mean()

2.5569542523794904

### Commentary

한 트위치 아이디가 보통 두 세가지의 직업을 하는 것으로 나타났다.